[QNNPACK, Sparsity] Add 8x1 block sparse kernels for aarch32.

kimishpatel opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #51120 [QNNPACK Sparsity] Add aarch64 kernel of 8x1 sparsity
* **#51119 [QNNPACK, Sparsity] Add 8x1 block sparse kernels for aarch32.**
* #51118 [QNNPACK, Sparsity] Code refactoring to allow for more generic block
* #50591 [QNNPACK, Sparsity] Add dyanmic linear sparse kernel for arm64
* #50590 [QNNPACK, Sparsity] Sparse kernel with 4x8 blocking
* #50589 [QNNPACK, Sparsity] Added prepacking base aarch32 kernels
* #50588 [QNNPACK, Sparsity] ARMV7, aarch32, kernels for dynamic linear

Summary:
Adds asm kernel for 8x1 block sparse kernel. Since ukernels is still
producing 4x8 blocks, similar to 1x4 sparsity pattern, we can use the
same prepacking kernel for activation. It does get a tiny bit hacky but
allows us to reuse the kernel.

Test Plan:
q8gemm-sparse-test
fully-connectest-sparse-test

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D26077765](https://our.internmc.facebook.com/intern/diff/D26077765)

kimishpatel committed:
[QNNPACK, Sparsity] Add 8x1 block sparse kernels for aarch32.

Summary:
Adds asm kernel for 8x1 block sparse kernel. Since ukernels is still
producing 4x8 blocks, similar to 1x4 sparsity pattern, we can use the
same prepacking kernel for activation. It does get a tiny bit hacky but
allows us to reuse the kernel.

Test Plan:
q8gemm-sparse-test
fully-connectest-sparse-test

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

kimishpatel committed:
Update on "[QNNPACK, Sparsity] Add 8x1 block sparse kernels for aarch32."


Summary:
Adds asm kernel for 8x1 block sparse kernel. Since ukernels is still
producing 4x8 blocks, similar to 1x4 sparsity pattern, we can use the
same prepacking kernel for activation. It does get a tiny bit hacky but
allows us to reuse the kernel.

Test Plan:
q8gemm-sparse-test
fully-connectest-sparse-test

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D26077765](https://our.internmc.facebook.com/intern/diff/D26077765)

[ghstack-poisoned]

kimishpatel committed:
Update on "[QNNPACK, Sparsity] Add 8x1 block sparse kernels for aarch32."


Summary:
Adds asm kernel for 8x1 block sparse kernel. Since ukernels is still
producing 4x8 blocks, similar to 1x4 sparsity pattern, we can use the
same prepacking kernel for activation. It does get a tiny bit hacky but
allows us to reuse the kernel.

Test Plan:
q8gemm-sparse-test
fully-connectest-sparse-test

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D26077765](https://our.internmc.facebook.com/intern/diff/D26077765)

[ghstack-poisoned]

kimishpatel committed:
Update on "[QNNPACK, Sparsity] Add 8x1 block sparse kernels for aarch32."


Summary:
Adds asm kernel for 8x1 block sparse kernel. Since ukernels is still
producing 4x8 blocks, similar to 1x4 sparsity pattern, we can use the
same prepacking kernel for activation. It does get a tiny bit hacky but
allows us to reuse the kernel.

Test Plan:
q8gemm-sparse-test
fully-connectest-sparse-test

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D26077765](https://our.internmc.facebook.com/intern/diff/D26077765)

[ghstack-poisoned]

kimishpatel committed:
Update on "[QNNPACK, Sparsity] Add 8x1 block sparse kernels for aarch32."


Summary:
Adds asm kernel for 8x1 block sparse kernel. Since ukernels is still
producing 4x8 blocks, similar to 1x4 sparsity pattern, we can use the
same prepacking kernel for activation. It does get a tiny bit hacky but
allows us to reuse the kernel.

Test Plan:
q8gemm-sparse-test
fully-connectest-sparse-test

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D26077765](https://our.internmc.facebook.com/intern/diff/D26077765)

[ghstack-poisoned]

kimishpatel committed:
Update on "[QNNPACK, Sparsity] Add 8x1 block sparse kernels for aarch32."


Summary:
Adds asm kernel for 8x1 block sparse kernel. Since ukernels is still
producing 4x8 blocks, similar to 1x4 sparsity pattern, we can use the
same prepacking kernel for activation. It does get a tiny bit hacky but
allows us to reuse the kernel.

Test Plan:
q8gemm-sparse-test
fully-connectest-sparse-test

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D26077765](https://our.internmc.facebook.com/intern/diff/D26077765)

[ghstack-poisoned]

kimishpatel committed:
Update on "[QNNPACK, Sparsity] Add 8x1 block sparse kernels for aarch32."


Summary:
Adds asm kernel for 8x1 block sparse kernel. Since ukernels is still
producing 4x8 blocks, similar to 1x4 sparsity pattern, we can use the
same prepacking kernel for activation. It does get a tiny bit hacky but
allows us to reuse the kernel.

Test Plan:
q8gemm-sparse-test
fully-connectest-sparse-test

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D26077765](https://our.internmc.facebook.com/intern/diff/D26077765)

[ghstack-poisoned]

AshkanAliabadi commented on aten/src/ATen/native/quantized/cpu/qnnpack/test/fully-connected-sparse.cc:
```diff
@@ -10,194 +10,135 @@
+
+SPARSE_OP_TEST(1, 4)
+#if CPUINFO_ARCH_ARM || CPUINFO_ARCH_ARM64
+SPARSE_OP_TEST(8, 1)
```
This will run for arm32 and arm64, but (1, 4) for all platforms, right?

kimishpatel replied:
Yes, because I did not implement (8, 1) pattern for SSE2.

kimishpatel committed:
Update on "[QNNPACK, Sparsity] Add 8x1 block sparse kernels for aarch32."


Summary:
Adds asm kernel for 8x1 block sparse kernel. Since ukernels is still
producing 4x8 blocks, similar to 1x4 sparsity pattern, we can use the
same prepacking kernel for activation. It does get a tiny bit hacky but
allows us to reuse the kernel.

Test Plan:
q8gemm-sparse-test
fully-connectest-sparse-test

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D26077765](https://our.internmc.facebook.com/intern/diff/D26077765)

[ghstack-poisoned]

kimishpatel committed:
Update on "[QNNPACK, Sparsity] Add 8x1 block sparse kernels for aarch32."


Summary:
Adds asm kernel for 8x1 block sparse kernel. Since ukernels is still
producing 4x8 blocks, similar to 1x4 sparsity pattern, we can use the
same prepacking kernel for activation. It does get a tiny bit hacky but
allows us to reuse the kernel.

Test Plan:
q8gemm-sparse-test
fully-connectest-sparse-test

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D26077765](https://our.internmc.facebook.com/intern/diff/D26077765)

[ghstack-poisoned]

kimishpatel commented on aten/src/ATen/native/quantized/cpu/qnnpack/src/operator-run.c:
```diff
@@ -1004,10 +1006,33 @@ enum pytorch_qnnp_status pytorch_qnnp_run_operator(
-      const uint32_t log2_mr = pytorch_qnnp_params.q8gemm_sparse.log2_mr;
-      const uint32_t nr = pytorch_qnnp_params.q8gemm_sparse.nr;
-      const uint32_t kr = pytorch_qnnp_params.q8gemm_sparse.kr;
+      uint32_t mr, log2_mr, nr, kr, log2_row_block_size;
```
I agree with you. Although all of those variables are value assigned by one of the ifs and else there is to error out. Just wanted to clarify. That said, I do see your reasoning about someone adding code and missing to assign value, but then what do you initialize it with? If you just do like -1, it is equally likely that we will introduce a bug that is hard to find. Only other thing I can imagine is to add validity check for each of those.

kimishpatel committed:
Update on "[QNNPACK, Sparsity] Add 8x1 block sparse kernels for aarch32."


Summary:
Adds asm kernel for 8x1 block sparse kernel. Since ukernels is still
producing 4x8 blocks, similar to 1x4 sparsity pattern, we can use the
same prepacking kernel for activation. It does get a tiny bit hacky but
allows us to reuse the kernel.

Test Plan:
q8gemm-sparse-test
fully-connectest-sparse-test

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D26077765](https://our.internmc.facebook.com/intern/diff/D26077765)

[ghstack-poisoned]

kimishpatel committed:
Update on "[QNNPACK, Sparsity] Add 8x1 block sparse kernels for aarch32."


Summary:
Adds asm kernel for 8x1 block sparse kernel. Since ukernels is still
producing 4x8 blocks, similar to 1x4 sparsity pattern, we can use the
same prepacking kernel for activation. It does get a tiny bit hacky but
allows us to reuse the kernel.

Test Plan:
q8gemm-sparse-test
fully-connectest-sparse-test

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D26077765](https://our.internmc.facebook.com/intern/diff/D26077765)

[ghstack-poisoned]

kimishpatel committed:
Update on "[QNNPACK, Sparsity] Add 8x1 block sparse kernels for aarch32."


Summary:
Adds asm kernel for 8x1 block sparse kernel. Since ukernels is still
producing 4x8 blocks, similar to 1x4 sparsity pattern, we can use the
same prepacking kernel for activation. It does get a tiny bit hacky but
allows us to reuse the kernel.

Test Plan:
q8gemm-sparse-test
fully-connectest-sparse-test

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D26077765](https://our.internmc.facebook.com/intern/diff/D26077765)

[ghstack-poisoned]

kimishpatel committed:
Update on "[QNNPACK, Sparsity] Add 8x1 block sparse kernels for aarch32."


Summary:
Adds asm kernel for 8x1 block sparse kernel. Since ukernels is still
producing 4x8 blocks, similar to 1x4 sparsity pattern, we can use the
same prepacking kernel for activation. It does get a tiny bit hacky but
allows us to reuse the kernel.

Test Plan:
q8gemm-sparse-test
fully-connectest-sparse-test

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D26077765](https://our.internmc.facebook.com/intern/diff/D26077765)

[ghstack-poisoned]

kimishpatel committed:
Update on "[QNNPACK, Sparsity] Add 8x1 block sparse kernels for aarch32."


Summary:
Adds asm kernel for 8x1 block sparse kernel. Since ukernels is still
producing 4x8 blocks, similar to 1x4 sparsity pattern, we can use the
same prepacking kernel for activation. It does get a tiny bit hacky but
allows us to reuse the kernel.

Test Plan:
q8gemm-sparse-test
fully-connectest-sparse-test

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D26077765](https://our.internmc.facebook.com/intern/diff/D26077765)

[ghstack-poisoned]

kimishpatel committed:
Update on "[QNNPACK, Sparsity] Add 8x1 block sparse kernels for aarch32."


Summary:
Adds asm kernel for 8x1 block sparse kernel. Since ukernels is still
producing 4x8 blocks, similar to 1x4 sparsity pattern, we can use the
same prepacking kernel for activation. It does get a tiny bit hacky but
allows us to reuse the kernel.

Test Plan:
q8gemm-sparse-test
fully-connectest-sparse-test

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D26077765](https://our.internmc.facebook.com/intern/diff/D26077765)

[ghstack-poisoned]

kimishpatel committed:
Update on "[QNNPACK, Sparsity] Add 8x1 block sparse kernels for aarch32."


Summary:
Adds asm kernel for 8x1 block sparse kernel. Since ukernels is still
producing 4x8 blocks, similar to 1x4 sparsity pattern, we can use the
same prepacking kernel for activation. It does get a tiny bit hacky but
allows us to reuse the kernel.

Test Plan:
q8gemm-sparse-test
fully-connectest-sparse-test

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D26077765](https://our.internmc.facebook.com/intern/diff/D26077765)

[ghstack-poisoned]

kimishpatel committed:
Update on "[QNNPACK, Sparsity] Add 8x1 block sparse kernels for aarch32."


Summary:
Adds asm kernel for 8x1 block sparse kernel. Since ukernels is still
producing 4x8 blocks, similar to 1x4 sparsity pattern, we can use the
same prepacking kernel for activation. It does get a tiny bit hacky but
allows us to reuse the kernel.

Test Plan:
q8gemm-sparse-test
fully-connectest-sparse-test

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D26077765](https://our.internmc.facebook.com/intern/diff/D26077765)

[ghstack-poisoned]
